File: benchmarks/05-factor-univariate/factor_univariate.py
#!/usr/bin/env python3
"""Experiment 05 -- Univariate polynomial factorisation (sympy column).

Same six kernels as ``factor_univariate.m``, same order.

The Mathilda side routes through FLINT when USE_FLINT is compiled in; sympy uses
its own pure-Python Zassenhaus.  Check values are FACTOR COUNTS: output ordering
is not canonical across systems, but the number of irreducible factors is.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import sympy

from harness import bench, check, require
from data import dense_upoly, factorable_upoly, sparse_upoly

require(["sympy", "sympy:factor", "sympy:factor_list", "sympy:gcd"])

x = sympy.Symbol("x")


def nfactors(e):
    """Match Length[FactorList[...]]: the unit term plus each distinct factor."""
    coeff, factors = sympy.factor_list(e)
    return len(factors) + 1


f1 = sympy.expand(factorable_upoly(8, x))
bench("Factor product of 8 quadratics", lambda: sympy.factor(f1))
check("Factor product of 8 quadratics", nfactors(f1))

f2 = sympy.expand(factorable_upoly(16, x))
bench("Factor product of 16 quadratics", lambda: sympy.factor(f2))
check("Factor product of 16 quadratics", nfactors(f2))

f3 = x ** 120 - 1
bench("Factor x^120 - 1", lambda: sympy.factor(f3))
check("Factor x^120 - 1", nfactors(f3))

f4 = sparse_upoly(60, x)
bench("Factor sparse degree 60", lambda: sympy.factor(f4))
check("Factor sparse degree 60", nfactors(f4))

f5 = dense_upoly(60, x)
bench("Factor dense degree 60", lambda: sympy.factor(f5))
check("Factor dense degree 60", nfactors(f5))

g1 = sympy.expand(factorable_upoly(10, x))
g2 = sympy.expand(factorable_upoly(6, x) * (x ** 4 + x + 1))
bench("PolynomialGCD of two deg-20 polys", lambda: sympy.gcd(g1, g2))
check("PolynomialGCD of two deg-20 polys",
      sympy.degree(sympy.gcd(g1, g2), x))

File: benchmarks/05-factor-univariate/factor_univariate.m
(* ==========================================================================
   Experiment 05 -- Univariate polynomial factorisation
   ==========================================================================
   WHAT IT MEASURES.  src/poly/facpoly.c and zupoly.c: Zassenhaus / Cantor-
   Zassenhaus over Z, and the FLINT bridge (src/poly/flint_bridge.c) when
   USE_FLINT is compiled in.

   READ THE BUILD WARNING IN REPORT.md BEFORE ACTING ON THESE ROWS.  Without
   FLINT >= 3.0 the makefile silently substitutes "the classical fallback", so
   these rows measure a different algorithm than an FLINT build does.  That is
   not a Mathilda defect; it is a build configuration, and the report says which
   one it measured.

   VALUE CHECKS ARE FACTOR COUNTS.  Factor's output ordering is not canonical
   across systems, but the NUMBER of irreducible factors (with multiplicity) is.
   ========================================================================== *)

Get["../harness.m"];
Get["../data.m"];

require[{"Factor", "FactorList", "Expand", "IrreduciblePolynomialQ",
         "PolynomialGCD", "Cyclotomic"}];

(* 1. A product of 8 distinct quadratics: the answer is known by construction. *)
f1 = Expand[factorableUPoly[8, x]];
bench["Factor product of 8 quadratics", Factor[f1];];
check["Factor product of 8 quadratics", Length[FactorList[f1]]];

(* 2. The same shape at 16 factors -- degree 32. *)
f2 = Expand[factorableUPoly[16, x]];
bench["Factor product of 16 quadratics", Factor[f2];];
check["Factor product of 16 quadratics", Length[FactorList[f2]]];

(* 3. x^n - 1: factors into cyclotomics.  Many small factors, high degree. *)
f3 = x^120 - 1;
bench["Factor x^120 - 1", Factor[f3];];
check["Factor x^120 - 1", Length[FactorList[f3]]];

(* 4. A sparse high-degree polynomial: cheap to store, expensive to factor. *)
f4 = sparseUPoly[60, x];
bench["Factor sparse degree 60", Factor[f4];];
check["Factor sparse degree 60", Length[FactorList[f4]]];

(* 5. A dense degree-60 polynomial with mild coefficients, probably irreducible:
   the WORST case, because the algorithm must exhaust its search to prove it. *)
f5 = denseUPoly[60, x];
bench["Factor dense degree 60", Factor[f5];];
check["Factor dense degree 60", Length[FactorList[f5]]];

(* 6. GCD of two large polynomials sharing a known factor. *)
g1 = Expand[factorableUPoly[10, x]];
g2 = Expand[factorableUPoly[6, x] (x^4 + x + 1)];
bench["PolynomialGCD of two deg-20 polys", PolynomialGCD[g1, g2];];
check["PolynomialGCD of two deg-20 polys", Exponent[PolynomialGCD[g1, g2], x]];
